src/compiler/checker.ts

```diff
@@ -4872,10 +4872,21 @@ namespace ts {
                     return Ternary.False;
                 }
 
-                let t = getReturnTypeOfSignature(target);
-                if (t === voidType) return result;
-                let s = getReturnTypeOfSignature(source);
-                return result & isRelatedTo(s, t, reportErrors);
+                let targetReturnType = getReturnTypeOfSignature(target);
+                if (targetReturnType === voidType) return result;
+                let sourceReturnType = getReturnTypeOfSignature(source);
+                
+                let targetReturnDecl = targetReturnType.symbol && getDeclarationOfKind(targetReturnType.symbol, SyntaxKind.ClassDeclaration);
+                let sourceReturnDecl = sourceReturnType.symbol && getDeclarationOfKind(sourceReturnType.symbol, SyntaxKind.ClassDeclaration);

```

Comment:
I'm not sure I understand the first suggestion. Could you clarify?

I moved the test out to `signaturesRelatedTo()` but It's kind of awkward to perform this exceptional check there and it's more LOC since we need to recompute some steps on the path to recovering the class declarations associated with the types. Do you think it's better now?